### DISCUSSION TITLE
wire:model on nested components

### DISCUSSION THREAD

**Discussion by mattgreenfield** [livewire/livewire#4139] [1. Help] [answered]
I've created a custom file input component that wraps up a HTML file input, a button and a list of files. I can use it like this: `<livewire:my-file-input />`.

Is there a way to add a `wire:model` to that and pass that to my components `<input type="file" />`?  e.g `<livewire:my-file-input wire:model="photo" />`.

Here's what I've thought might work but didn't:
1.  From what I understand, `wire:model` is listening for `input` events. In Vue I can emit an `input` event from anywhere but it doesn't seem to work in livewire. Besides, I'd have to add a change listener to the input and that seem to go against the whole ethos of using model.
2. Something like how blade does `{{ $attributes }}` so that `<input type="file" {{ $attributes }} />` would work but I understand livewire doesn't register things in child component HTML like that.

Thanks for your help,

**Comment by mattkibbler** [comment, marked as answer]
Hey @mattgreenfield,

I would recommend sponsoring @calebporzio for a month so you get access to the Livewire screencasts and project repo, they're great and include, amongst many other things, handling a file upload like this.

In short, your file input wouldn't necessarily be a Livewire component, it can be a simply blade component that is included within a Livewire component.

Let's say you have a profile page with an avatar upload. You could have a ProfileForm Livewire component, within the template for that Livewire component you can include your file upload blade component and wire that up to a property on your ProfileForm Livewire component.

Something like:

**resources/views/livewire/profile-form.blade.php**

```html
<form>
    <x-file-input wire:model="avatar" />
    <button type="button" wire:click="save">Save</button>
</form>
```
**resources/views/components/file-input.blade.php**
```html
<input type="file" {{$attributes}} />
```

**app/Http/Livewire/ProfileForm.php**

```php
<?php

namespace App\Http\Livewire;

use Livewire\Component;
use Livewire\WithFileUploads;

class ProfileForm extends Component
{

    use WithFileUploads;

    public $avatar;

    public function save() {
        dd($this->avatar);
    }

    public function render()
    {
        return view('livewire.profile-form');
    }
}

```

I hope that helps, let me know if you have any problems.

**Reply by mattgreenfield** [reply]
Thank you, this is great. I'm new to Blade, Alpine and Livewire and I think the hardest part has been knowing when to use each one and how they interact. I'll take a look at sponsorship too.

It was my understanding that Alpine runs in the browser and so 'sees' html, meaning you can compose attributes like this - they don't have to be in the same file, just in the same HTML page and scope.
It's news to me that livewire works the same way and the PHP files don't have to 'know' about the methods and attributes.

Thank you again!

**Reply by mattkibbler** [reply]
No worries!

Hopefully this helps clear things up...

- If you need to add a bit of dynamic interaction or reactivity, like opening a dropdown menu or modal, use Alpine. As it says on the Alpine site; "think of it like jQuery for the modern web".
- If you need to add dynamic interaction or reactivity that involves the backend in some way (things like submitting form data, managing database records etc) then use a Livewire component to act as a sort of controller for the reactivity.
- You'll use Blade regardless to render HTML from your Laravel app, with or without Livewire and Alpine.

Livewire and Alpine are separate, but play nicely together. You're not actually using Alpine in your example there, just Livewire.
